### ISSUE TITLE
[core] Provide custom PmdReporter for RuleSetLoader

### ISSUE THREAD

**Issue by adangel** [pmd/pmd#6126]
**Is your feature request related to a problem? Please describe.**
When using the [RuleSetLoader](https://github.com/pmd/pmd/blob/main/pmd-core/src/main/java/net/sourceforge/pmd/lang/rule/RuleSetLoader.java) with the default configuration (via default ctor), the reporter is set to quiet. And there is no way to set the reporter (method `withReporter`) is package private.

I myself run into this issue when creating a unit test to verify a ruleset doesn't have deprecated rule by simply looking at what the reporter reports - ideally, it should not report any warnings when loading a ruleset. See https://github.com/pmd/build-tools/blob/f18951a4d5b4ca99dd060f6e76d72d340a9111df/src/test/java/net/sourceforge/pmd/buildtools/LoadRulesetTest.java#L38-L49 . It uses the PMDConfiguration which by default uses a logger (https://github.com/pmd/pmd/blob/2a719ddcb874060c4338b179a766221ba2aa34d4/pmd-core/src/main/java/net/sourceforge/pmd/PMDConfiguration.java#L119).

Similar use case is now in the sonar-pmd plugin: https://github.com/jborgers/sonar-pmd/blob/2397557821de0a5d86b4cf26d0787a88cd930b53/sonar-pmd-plugin/src/main/java/org/sonar/plugins/pmd/AbstractPmdExecutor.java#L206-L221

There seems to be some opportunity for improvement around this API and logging/reporting.

**Describe the solution you'd like**
The easiest way would be to just make the method `withReporter` public. But maybe there were good reasons not to do this.

**Describe alternatives you've considered**
Maybe deprecating the default ctor of RuleSetLoader, so that you always need to use the factory method `RuleSetLoader.fromPmdConfig()`:

```java
    private static RuleSet parseRuleSetWithReporter(PmdReporter reporter, String ruleSetFilePath) {
        PMDConfiguration pmdConfiguration = new PMDConfiguration();
        pmdConfiguration.setReporter(reporter);
        RuleSetLoader loader = RuleSetLoader.fromPmdConfig(pmdConfiguration);
        return loader.loadFromResource(ruleSetFilePath);
    }
```

**Additional context**
Maybe related older issues are:
* #3816
* #897

**Comment by adangel** [comment]
I guess, the main issue is, that the following is surprising:

```java
RuleSetLoader loader = new RuleSetLoader().warnDeprecation(true);
loader.loadFromResource(ruleSetFilePath)
```

And - despite the warn deprecation setting - you won't see any deprecation warnings, because the default reporter is quiet.
